File: php/Project_PHP/app/Models/ArvoreBinaria/arvoreBinaria.php
<?php
namespace Estruturas;

use Estruturas\NoArvore as No;

class ArvoreBinaria
{
    private ?No $root = null;

    public function __construct()
    {
    }
    
    public function taCompleto(No $no): bool
    {
        return $no->getEsquerdo() !== null && $no->getDireito() !== null;
    }

    public function inserir(No $valor): void
    {
        if ($this->root === null) {
            $this->root = $valor;
        } else {
            $this->_inserir($this->root, $valor);
        }
    }

    private function _inserir(No $pai, No $valor): void
    {
        if ($pai->getEsquerdo() === null) {
            $pai->setEsquerdo($valor);
            return;
        }
        
        elseif ($pai->getDireito() === null) {
            $pai->setDireito($valor);
            return;
        }
        
        else {
            $this->_inserir($pai->getEsquerdo(), $valor);

            if ($this->taCompleto($pai->getEsquerdo())) {
                $this->_inserir($pai->getDireito(), $valor);
                return;
            } else {
                $this->_inserir($pai->getEsquerdo(), $valor);
                return;
            }
        }
    }

    public function exibir(): void
    {
        if ($this->root !== null) {
            $this->_exibir($this->root, 0, "R: ");
        }
    }

    private function _exibir(No $atual, int $nivel, string $prefixo): void
    {
        $indentacao = str_repeat(" ", $nivel);
        echo $indentacao . $prefixo . $atual->getValor() . "\n";

        if ($atual->getEsquerdo() !== null) {
            $this->_exibir($atual->getEsquerdo(), $nivel + 2, "E: ");
        }
        
        if ($atual->getDireito() !== null) {
            $this->_exibir($atual->getDireito(), $nivel + 2, "D: ");
        }
    }
}

File: python/src/ArvoreBinaria/arvoreBinaria.py
from src.No.node import No


class ArvoreBinaria:
    def __init__(self):
        self.raiz: No = None
    
    def inserir(self, valor: No) -> None:
        if not self.raiz:
            self.raiz = valor
        else:
            self._inserir(self.raiz, valor)
    
    def _inserir(self, pai: No, valor: No) -> None:
        if not pai.esquerdo:
            pai.esquerdo = valor
            return
        elif not pai.direito:
            pai.direito = valor
            return
        else:
            if self._altura(pai.esquerdo) <= self._altura(pai.direito):
                return self._inserir(pai.esquerdo, valor)
            else:
                return self._inserir(pai.direito, valor)

    def taCompleto(self, no: No) -> bool:
        return no.esquerdo and no.direito
    
    def exibir(self):
        self._exibir(self.raiz, 0, "R: ")

    def _exibir(self, atual: No, nivel: int, prefixo: str) -> None:
        print(" " * nivel, prefixo, atual.valor)
        if atual.esquerdo:
            self._exibir(atual.esquerdo, nivel + 2, "E: ")
        if atual.direito:
            self._exibir(atual.direito, nivel + 2, "D: ")
    
    def altura(self):
        return self._altura(self.raiz)

    def _altura(self, no: No):
        if no is None:
            return 0
        else: 
            return 1 + max(self._altura(no.esquerdo), self._altura(no.direito))
        


